File: ts/src/anomaly/detectors.ts
/**
 * Anomaly detectors
 */

/**
 * Result of anomaly detection
 */
export interface AnomalyResult {
  /** Boolean array indicating anomalies */
  isAnomaly: boolean[];
  /** Anomaly scores for each point */
  scores: number[];
  /** Threshold used for detection */
  threshold: number;
  /** Indices of anomalies */
  anomalyIndices: number[];
}

/**
 * Interface for anomaly detectors
 */
export interface AnomalyDetector {
  /** Name of the detector */
  readonly name: string;

  /** Fit the detector to training data */
  fit(data: number[]): void;

  /** Detect anomalies in data */
  detect(data: number[]): AnomalyResult;

  /** Get anomaly scores (higher = more anomalous) */
  score(data: number[]): number[];
}

/**
 * Z-Score based anomaly detector
 */
export class ZScoreDetector implements AnomalyDetector {
  readonly name = 'Z-Score';
  private threshold: number;
  private mean = 0;
  private std = 1;
  private fitted = false;

  constructor(threshold: number = 3.0) {
    this.threshold = threshold;
  }

  fit(data: number[]): void {
    const n = data.length;
    this.mean = data.reduce((a, b) => a + b, 0) / n;
    const variance = data.reduce((sum, x) => sum + (x - this.mean) ** 2, 0) / n;
    this.std = Math.sqrt(variance);
    this.fitted = true;
  }

  score(data: number[]): number[] {
    if (this.std === 0) {
      return data.map(() => 0);
    }
    return data.map((x) => Math.abs((x - this.mean) / this.std));
  }

  detect(data: number[]): AnomalyResult {
    if (!this.fitted) {
      this.fit(data);
    }

    const scores = this.score(data);
    const isAnomaly = scores.map((s) => s > this.threshold);
    const anomalyIndices = isAnomaly
      .map((a, i) => (a ? i : -1))
      .filter((i) => i >= 0);

    return {
      isAnomaly,
      scores,
      threshold: this.threshold,
      anomalyIndices,
    };
  }
}

/**
 * Interquartile Range (IQR) based anomaly detector
 */
export class IQRDetector implements AnomalyDetector {
  readonly name = 'IQR';
  private multiplier: number;
  private q1 = 0;
  private q3 = 0;
  private iqr = 0;
  private fitted = false;

  constructor(multiplier: number = 1.5) {
    this.multiplier = multiplier;
  }

  fit(data: number[]): void {
    const sorted = [...data].sort((a, b) => a - b);
    const n = sorted.length;
    this.q1 = sorted[Math.floor(n * 0.25)];
    this.q3 = sorted[Math.floor(n * 0.75)];
    this.iqr = this.q3 - this.q1;
    this.fitted = true;
  }

  score(data: number[]): number[] {
    if (this.iqr === 0) {
      return data.map(() => 0);
    }
    const lower = this.q1 - this.multiplier * this.iqr;
    const upper = this.q3 + this.multiplier * this.iqr;

    return data.map((x) => {
      if (x < lower) return (lower - x) / this.iqr;
      if (x > upper) return (x - upper) / this.iqr;
      return 0;
    });
  }

  detect(data: number[]): AnomalyResult {
    if (!this.fitted) {
      this.fit(data);
    }

    const lower = this.q1 - this.multiplier * this.iqr;
    const upper = this.q3 + this.multiplier * this.iqr;

    const isAnomaly = data.map((x) => x < lower || x > upper);
    const scores = this.score(data);
    const anomalyIndices = isAnomaly
      .map((a, i) => (a ? i : -1))
      .filter((i) => i >= 0);

    return {
      isAnomaly,
      scores,
      threshold: this.multiplier,
      anomalyIndices,
    };
  }
}

/**
 * Median Absolute Deviation (MAD) based anomaly detector
 * More robust to outliers than Z-score
 */
export class MADDetector implements AnomalyDetector {
  readonly name = 'MAD';
  private threshold: number;
  private median = 0;
  private mad = 0;
  private fitted = false;

  constructor(threshold: number = 3.5) {
    this.threshold = threshold;
  }

  private calculateMedian(data: number[]): number {
    const sorted = [...data].sort((a, b) => a - b);
    const mid = Math.floor(sorted.length / 2);
    return sorted.length % 2 !== 0
      ? sorted[mid]
      : (sorted[mid - 1] + sorted[mid]) / 2;
  }

  fit(data: number[]): void {
    this.median = this.calculateMedian(data);
    const deviations = data.map((x) => Math.abs(x - this.median));
    this.mad = this.calculateMedian(deviations);
    this.fitted = true;
  }

  score(data: number[]): number[] {
    // Modified Z-score using MAD
    // Constant 0.6745 makes MAD consistent with standard deviation for normal distribution
    const k = 0.6745;
    if (this.mad === 0) {
      return data.map(() => 0);
    }
    return data.map((x) => Math.abs((x - this.median) / (this.mad / k)));
  }

  detect(data: number[]): AnomalyResult {
    if (!this.fitted) {
      this.fit(data);
    }

    const scores = this.score(data);
    const isAnomaly = scores.map((s) => s > this.threshold);
    const anomalyIndices = isAnomaly
      .map((a, i) => (a ? i : -1))
      .filter((i) => i >= 0);

    return {
      isAnomaly,
      scores,
      threshold: this.threshold,
      anomalyIndices,
    };
  }
}

File: crates/rustful-anomaly/src/detectors.rs
//! Anomaly detectors

use rustful_core::Result;
use serde::{Deserialize, Serialize};

/// Anomaly detection result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AnomalyResult {
    pub is_anomaly: Vec<bool>,
    pub scores: Vec<f64>,
    pub threshold: f64,
}

/// Anomaly detector trait
pub trait AnomalyDetector {
    fn fit(&mut self, data: &[f64]) -> Result<()>;
    fn detect(&self, data: &[f64]) -> Result<AnomalyResult>;
    fn score(&self, data: &[f64]) -> Result<Vec<f64>>;
}

/// Z-Score based anomaly detector
#[derive(Debug, Clone)]
pub struct ZScoreDetector {
    threshold: f64,
    mean: f64,
    std_dev: f64,
    fitted: bool,
}

impl ZScoreDetector {
    pub fn new(threshold: f64) -> Self {
        Self {
            threshold,
            mean: 0.0,
            std_dev: 1.0,
            fitted: false,
        }
    }
}

impl Default for ZScoreDetector {
    fn default() -> Self {
        Self::new(3.0)
    }
}

impl AnomalyDetector for ZScoreDetector {
    fn fit(&mut self, data: &[f64]) -> Result<()> {
        let n = data.len() as f64;
        self.mean = data.iter().sum::<f64>() / n;
        self.std_dev = (data.iter().map(|x| (x - self.mean).powi(2)).sum::<f64>() / n).sqrt();
        self.fitted = true;
        Ok(())
    }

    fn detect(&self, data: &[f64]) -> Result<AnomalyResult> {
        let scores = self.score(data)?;
        let is_anomaly = scores.iter().map(|&s| s.abs() > self.threshold).collect();
        Ok(AnomalyResult {
            is_anomaly,
            scores,
            threshold: self.threshold,
        })
    }

    fn score(&self, data: &[f64]) -> Result<Vec<f64>> {
        if self.std_dev == 0.0 {
            return Ok(vec![0.0; data.len()]);
        }
        Ok(data.iter().map(|&x| (x - self.mean) / self.std_dev).collect())
    }
}

/// IQR-based anomaly detector
#[derive(Debug, Clone)]
pub struct IQRDetector {
    multiplier: f64,
    q1: f64,
    q3: f64,
    fitted: bool,
}

impl IQRDetector {
    pub fn new(multiplier: f64) -> Self {
        Self {
            multiplier,
            q1: 0.0,
            q3: 0.0,
            fitted: false,
        }
    }
}

impl Default for IQRDetector {
    fn default() -> Self {
        Self::new(1.5)
    }
}

impl AnomalyDetector for IQRDetector {
    fn fit(&mut self, data: &[f64]) -> Result<()> {
        let mut sorted = data.to_vec();
        sorted.sort_by(|a, b| a.partial_cmp(b).unwrap());
        let n = sorted.len();
        self.q1 = sorted[n / 4];
        self.q3 = sorted[3 * n / 4];
        self.fitted = true;
        Ok(())
    }

    fn detect(&self, data: &[f64]) -> Result<AnomalyResult> {
        let iqr = self.q3 - self.q1;
        let lower = self.q1 - self.multiplier * iqr;
        let upper = self.q3 + self.multiplier * iqr;

        let is_anomaly: Vec<bool> = data.iter().map(|&x| x < lower || x > upper).collect();
        let scores: Vec<f64> = data
            .iter()
            .map(|&x| {
                if x < lower {
                    (lower - x) / iqr
                } else if x > upper {
                    (x - upper) / iqr
                } else {
                    0.0
                }
            })
            .collect();

        Ok(AnomalyResult {
            is_anomaly,
            scores,
            threshold: self.multiplier,
        })
    }

    fn score(&self, data: &[f64]) -> Result<Vec<f64>> {
        let iqr = self.q3 - self.q1;
        Ok(data
            .iter()
            .map(|&x| {
                if iqr == 0.0 {
                    0.0
                } else {
                    ((x - self.q1) / iqr).abs()
                }
            })
            .collect())
    }
}
